# 3. Pandas

Pandas is a Python library used for working with data sets. It has functions for analyzing, cleaning, exploring, and manipulating data.

Pandas allows us to analyze big data and make conclusions based on statistical theories. Pandas can clean messy data sets, and make them readable and relevant. Relevant data is very important in data science.

When working with **tabular or structured data**, we can do following things easily:

- Import data
- Clean up messy data
- Explore data, gain insight into data
- Process and prepare your data for analysis
- Analyse your data (together with scikit-learn, statsmodels, ...)

## 3.1. How to Use Pandas

Do you remember how we used **numpy** library? We **imported** the numpy using `import numpy as np`. We will do simliar thing with Pandas.

To access **Pandas** and its functions import it in your Python code like this:

In [2]:
import numpy as np

In [3]:
import pandas as pd

We shorten the imported name to `pd` for better readability of code using **Pandas**. This is a widely adopted convention that makes your code more readable for everyone working on it. We recommend to always use `import pandas as pd`.

## 3.2. Data structures in Pandas

We will learn about two data structures provided by Pandas, `Series` and `DataFrame`

### 3.2.1 Series

`Series` is a one-dimensional labeled array capable of holding any data type (integers, strings, floating point numbers, Python objects, etc.). The axis labels are collectively referred to as the **index**. 

You can think of each `Series` as **one column of excel sheet**.

The basic method to create a `Series` is to call:

In [4]:
s = pd.Series([1, 3, 5, 4, 6, 8])
print(s)

0    1
1    3
2    5
3    4
4    6
5    8
dtype: int64


Above, the fist column is index. Since we did not defined the index while creating a series `s`, `s` has a default index, which starts from 0.

The second column is the values in `s`. We gave the list ([1, 3, 5, 4, 6, 8]) as the input to create `s`, so the values in the list ([1, 3, 5, 4, 6, 8]) become the values.

Since every value in `s` are integers, the datatype of `s` becomes `int64`.

In [5]:
s = pd.Series([1, 3.5, 5, 4, 6.2, 8])
print(s)

0    1.0
1    3.5
2    5.0
3    4.0
4    6.2
5    8.0
dtype: float64


When there are both integers and floating points numbers inside the list given to `s`, the datatype of `s` becomes `float64`.

We can also define the index of `Series` like this:

In [6]:
s = pd.Series([1, 3.5, 5, 4, 6.2, 8], index=['a', 'b', 'c', 'd', 'e', 'f'])
print(s)

a    1.0
b    3.5
c    5.0
d    4.0
e    6.2
f    8.0
dtype: float64


The `index` keyword in `pd.Series(,index=)` defines what the index of Series will be.

### 3.2.2. DataFrame

While `Series` is a one-dimensional labeled array, `DataFrame` is a two-dimensional data structure that holds data likea two-dimension array or a table with rows and columns.

Therefore, `DataFrame` is just like the each excel sheet. You will mostly use `DataFrame` when analyzing data.

You can create a `DataFrame` using `dictionary` like follows:

In [10]:
df = pd.DataFrame({'Apples': [1,3,7,4],
                  'Bananas': [2,6,3,5]})

In [12]:
df

Apples  Bananas
0       1        2
1       3        6
2       7        3
3       4        5

To summarize, `DataFrame` is like the excel sheet containing multiple columns, and each columns in `DataFrame` is `Series`.

![](https://codetorial.net/pandas/_images/object_creation_01.png)

## 3.3. Reading csv file using Pandas

csv file is the excel file, which is easy to read and write using Python (Pandas). You can read the contents in csv file and create it as the `DataFrame` using Pandas.

For the rest of the tutorial, we will use the csv file named `titanic.csv`. You have to place it in the same directory with this `.ipynb` file. (ex. If this `lab_3.ipynb` file is located in `C:/Documents`, `titanic.csv` should also be in `C:/Documents` too.)

Then, we can read the csv file like follows:

In [15]:
df = pd.read_csv("titanic.csv")

In [16]:
df

PassengerId  Survived  Pclass  \
0              1         0       3   
1              2         1       1   
2              3         1       3   
3              4         1       1   
4              5         0       3   
..           ...       ...     ...   
886          887         0       2   
887          888         1       1   
888          889         0       3   
889          890         1       1   
890          891         0       3   

                                                  Name     Sex   Age  SibSp  \
0                              Braund, Mr. Owen Harris    male  22.0      1   
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                               Heikkinen, Miss. Laina  female  26.0      0   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                             Allen, Mr. William Henry    male  35.0      0   
..                                                 ...     ...   ...    ...   
886                              Montvila, Rev. Juozas    male  27.0      0   
887                       Graham, Miss. Margaret Edith  female  19.0      0   
888           Johnston, Miss. Catherine Helen "Carrie"  female   NaN      1   
889                              Behr, Mr. Karl Howell    male  26.0      0   
890                                Dooley, Mr. Patrick    male  32.0      0   

     Parch            Ticket     Fare Cabin Embarked  
0        0         A/5 21171   7.2500   NaN        S  
1        0          PC 17599  71.2833   C85        C  
2        0  STON/O2. 3101282   7.9250   NaN        S  
3        0            113803  53.1000  C123        S  
4        0            373450   8.0500   NaN        S  
..     ...               ...      ...   ...      ...  
886      0            211536  13.0000   NaN        S  
887      0            112053  30.0000   B42        S  
888      2        W./C. 6607  23.4500   NaN        S  
889      0            111369  30.0000  C148        C  
890      0            370376   7.7500   NaN        Q  

[891 rows x 12 columns]

There are 12 columns. If we want to set the `PassengerId` column as the index of `DataFrame`, we can use `index_col` like follows:

In [17]:
df = pd.read_csv("titanic.csv", index_col=0)

In [18]:
df

Survived  Pclass  \
PassengerId                     
1                   0       3   
2                   1       1   
3                   1       3   
4                   1       1   
5                   0       3   
...               ...     ...   
887                 0       2   
888                 1       1   
889                 0       3   
890                 1       1   
891                 0       3   

                                                          Name     Sex   Age  \
PassengerId                                                                    
1                                      Braund, Mr. Owen Harris    male  22.0   
2            Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0   
3                                       Heikkinen, Miss. Laina  female  26.0   
4                 Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0   
5                                     Allen, Mr. William Henry    male  35.0   
...                                                        ...     ...   ...   
887                                      Montvila, Rev. Juozas    male  27.0   
888                               Graham, Miss. Margaret Edith  female  19.0   
889                   Johnston, Miss. Catherine Helen "Carrie"  female   NaN   
890                                      Behr, Mr. Karl Howell    male  26.0   
891                                        Dooley, Mr. Patrick    male  32.0   

             SibSp  Parch            Ticket     Fare Cabin Embarked  
PassengerId                                                          
1                1      0         A/5 21171   7.2500   NaN        S  
2                1      0          PC 17599  71.2833   C85        C  
3                0      0  STON/O2. 3101282   7.9250   NaN        S  
4                1      0            113803  53.1000  C123        S  
5                0      0            373450   8.0500   NaN        S  
...            ...    ...               ...      ...   ...      ...  
887              0      0            211536  13.0000   NaN        S  
888              0      0            112053  30.0000   B42        S  
889              1      2        W./C. 6607  23.4500   NaN        S  
890              0      0            111369  30.0000  C148        C  
891              0      0            370376   7.7500   NaN        Q  

[891 rows x 11 columns]

`index_col=0` means that we will set the first (always remember. In Python, 0 is the first) column as `index` of `DataFrame`.

We can check the `index` using `.index`

In [19]:
df.index

Int64Index([  1,   2,   3,   4,   5,   6,   7,   8,   9,  10,
            ...
            882, 883, 884, 885, 886, 887, 888, 889, 890, 891],
           dtype='int64', name='PassengerId', length=891)

## 3.4. Viewing Data

We will learn about various methods of viewing data in `Pandas` data structures.

### 3.4.1. df.head() and df.tail()

To view the top and bottoms rows of the dataframe, we can use `df.head()` and `df.tail()`, respectively.

In [20]:
df.head()

Survived  Pclass  \
PassengerId                     
1                   0       3   
2                   1       1   
3                   1       3   
4                   1       1   
5                   0       3   

                                                          Name     Sex   Age  \
PassengerId                                                                    
1                                      Braund, Mr. Owen Harris    male  22.0   
2            Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0   
3                                       Heikkinen, Miss. Laina  female  26.0   
4                 Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0   
5                                     Allen, Mr. William Henry    male  35.0   

             SibSp  Parch            Ticket     Fare Cabin Embarked  
PassengerId                                                          
1                1      0         A/5 21171   7.2500   NaN        S  
2                1      0          PC 17599  71.2833   C85        C  
3                0      0  STON/O2. 3101282   7.9250   NaN        S  
4                1      0            113803  53.1000  C123        S  
5                0      0            373450   8.0500   NaN        S

In [21]:
df.tail()

Survived  Pclass                                      Name  \
PassengerId                                                               
887                 0       2                     Montvila, Rev. Juozas   
888                 1       1              Graham, Miss. Margaret Edith   
889                 0       3  Johnston, Miss. Catherine Helen "Carrie"   
890                 1       1                     Behr, Mr. Karl Howell   
891                 0       3                       Dooley, Mr. Patrick   

                Sex   Age  SibSp  Parch      Ticket   Fare Cabin Embarked  
PassengerId                                                                
887            male  27.0      0      0      211536  13.00   NaN        S  
888          female  19.0      0      0      112053  30.00   B42        S  
889          female   NaN      1      2  W./C. 6607  23.45   NaN        S  
890            male  26.0      0      0      111369  30.00  C148        C  
891            male  32.0      0      0      370376   7.75   NaN        Q

If we want to see top 3 rows, we can use `df.head(3)`

In [24]:
df.head(3)

Survived  Pclass  \
PassengerId                     
1                   0       3   
2                   1       1   
3                   1       3   

                                                          Name     Sex   Age  \
PassengerId                                                                    
1                                      Braund, Mr. Owen Harris    male  22.0   
2            Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0   
3                                       Heikkinen, Miss. Laina  female  26.0   

             SibSp  Parch            Ticket     Fare Cabin Embarked  
PassengerId                                                          
1                1      0         A/5 21171   7.2500   NaN        S  
2                1      0          PC 17599  71.2833   C85        C  
3                0      0  STON/O2. 3101282   7.9250   NaN        S

### 3.4.2. df.index and df.columns

We can check the index and column names using `df.index` and `df.columns`.

In [25]:
df.index

Int64Index([  1,   2,   3,   4,   5,   6,   7,   8,   9,  10,
            ...
            882, 883, 884, 885, 886, 887, 888, 889, 890, 891],
           dtype='int64', name='PassengerId', length=891)

In [26]:
df.columns

Index(['Survived', 'Pclass', 'Name', 'Sex', 'Age', 'SibSp', 'Parch', 'Ticket',
       'Fare', 'Cabin', 'Embarked'],
      dtype='object')

### 3.4.3. df.to_numpy()

`Numpy` and `Pandas` are very compatible libraries. We can change `DataFrame` datatype to `Numpy array` datatype using `df.to_numpy()`.

In [27]:
df.to_numpy()

array([[0, 3, 'Braund, Mr. Owen Harris', ..., 7.25, nan, 'S'],
       [1, 1, 'Cumings, Mrs. John Bradley (Florence Briggs Thayer)', ...,
        71.2833, 'C85', 'C'],
       [1, 3, 'Heikkinen, Miss. Laina', ..., 7.925, nan, 'S'],
       ...,
       [0, 3, 'Johnston, Miss. Catherine Helen "Carrie"', ..., 23.45,
        nan, 'S'],
       [1, 1, 'Behr, Mr. Karl Howell', ..., 30.0, 'C148', 'C'],
       [0, 3, 'Dooley, Mr. Patrick', ..., 7.75, nan, 'Q']], dtype=object)

### 3.4.4. df.describe()

`df.describe` shows a quick statistic summary of data. (only numerical columns)

In [28]:
df.describe()

Survived      Pclass         Age       SibSp       Parch        Fare
count  891.000000  891.000000  714.000000  891.000000  891.000000  891.000000
mean     0.383838    2.308642   29.699118    0.523008    0.381594   32.204208
std      0.486592    0.836071   14.526497    1.102743    0.806057   49.693429
min      0.000000    1.000000    0.420000    0.000000    0.000000    0.000000
25%      0.000000    2.000000   20.125000    0.000000    0.000000    7.910400
50%      0.000000    3.000000   28.000000    0.000000    0.000000   14.454200
75%      1.000000    3.000000   38.000000    1.000000    0.000000   31.000000
max      1.000000    3.000000   80.000000    8.000000    6.000000  512.329200

## 3.5. Selecting Data

We will learn about how to select data in specific locaiton, or by condition in `DataFrame`.

### 3.5.1. Selecting Column

If we want to select column named **Age**, we can use `df.Age` or `df['Age']`.

In [29]:
df.Age

PassengerId
1      22.0
2      38.0
3      26.0
4      35.0
5      35.0
       ... 
887    27.0
888    19.0
889     NaN
890    26.0
891    32.0
Name: Age, Length: 891, dtype: float64

In [30]:
df['Age']

PassengerId
1      22.0
2      38.0
3      26.0
4      35.0
5      35.0
       ... 
887    27.0
888    19.0
889     NaN
890    26.0
891    32.0
Name: Age, Length: 891, dtype: float64

### 3.5.2. Selecting Row

We can select row with specific index using `df.loc`.

For example, when selecting row with `PassengerId=3`, we can do it as follows:

In [33]:
df.loc[3]

Survived                         1
Pclass                           3
Name        Heikkinen, Miss. Laina
Sex                         female
Age                           26.0
SibSp                            0
Parch                            0
Ticket            STON/O2. 3101282
Fare                         7.925
Cabin                          NaN
Embarked                         S
Name: 3, dtype: object

We can also select multiple rows.

For example, when selecting row with `PassengerId=3 or 6`, we can do it as follows:

In [36]:
df.loc[[3,6],:]

Survived  Pclass                    Name     Sex   Age  SibSp  \
PassengerId                                                                  
3                   1       3  Heikkinen, Miss. Laina  female  26.0      0   
6                   0       3        Moran, Mr. James    male   NaN      0   

             Parch            Ticket    Fare Cabin Embarked  
PassengerId                                                  
3                0  STON/O2. 3101282  7.9250   NaN        S  
6                0            330877  8.4583   NaN        Q

If we want to select rows from `PassengerId=3 ~ 6`, we can do it as follows:

In [37]:
df.loc[3:6,:]

Survived  Pclass                                          Name  \
PassengerId                                                                   
3                   1       3                        Heikkinen, Miss. Laina   
4                   1       1  Futrelle, Mrs. Jacques Heath (Lily May Peel)   
5                   0       3                      Allen, Mr. William Henry   
6                   0       3                              Moran, Mr. James   

                Sex   Age  SibSp  Parch            Ticket     Fare Cabin  \
PassengerId                                                                
3            female  26.0      0      0  STON/O2. 3101282   7.9250   NaN   
4            female  35.0      1      0            113803  53.1000  C123   
5              male  35.0      0      0            373450   8.0500   NaN   
6              male   NaN      0      0            330877   8.4583   NaN   

            Embarked  
PassengerId           
3                  S  
4                  S  
5                  S  
6                  Q

If we want to select row with `PassengerId=3` and get value for column `Name`, we can do as follows:

In [38]:
df.loc[3,'Name']

'Heikkinen, Miss. Laina'

### 3.5.3. Selecting by index location

Even though `DataFrame` has index and we selected row based on index value using `df.loc`, we can also select row using the position of row (which means, the first row becomes position `0`) using `df.iloc`.

If we want to select `n`-th row, we can use `df.iloc[n-1]`.

In [41]:
df.iloc[2]

Survived                         1
Pclass                           3
Name        Heikkinen, Miss. Laina
Sex                         female
Age                           26.0
SibSp                            0
Parch                            0
Ticket            STON/O2. 3101282
Fare                         7.925
Cabin                          NaN
Embarked                         S
Name: 3, dtype: object

It also works for columns. When we want to select `n`-th column, we can use `df.iloc[:,n-1]`.

In [42]:
df.iloc[:,2]

PassengerId
1                                Braund, Mr. Owen Harris
2      Cumings, Mrs. John Bradley (Florence Briggs Th...
3                                 Heikkinen, Miss. Laina
4           Futrelle, Mrs. Jacques Heath (Lily May Peel)
5                               Allen, Mr. William Henry
                             ...                        
887                                Montvila, Rev. Juozas
888                         Graham, Miss. Margaret Edith
889             Johnston, Miss. Catherine Helen "Carrie"
890                                Behr, Mr. Karl Howell
891                                  Dooley, Mr. Patrick
Name: Name, Length: 891, dtype: object

In [43]:
df.iloc[3:6,2:4]

Name     Sex
PassengerId                                                      
4            Futrelle, Mrs. Jacques Heath (Lily May Peel)  female
5                                Allen, Mr. William Henry    male
6                                        Moran, Mr. James    male

In [44]:
df.iloc[[1,2,4],[0,2]]

Survived                                               Name
PassengerId                                                             
2                   1  Cumings, Mrs. John Bradley (Florence Briggs Th...
3                   1                             Heikkinen, Miss. Laina
5                   0                           Allen, Mr. William Henry

In [46]:
df.iloc[0,2]

'Braund, Mr. Owen Harris'

### 3.5.4. Selecting based on condition

We can select rows based on the condition of column.

For example, we can select dataframe which `Survived` column is `1` as follows:

In [48]:
df[df.Survived == 1]

Survived  Pclass  \
PassengerId                     
2                   1       1   
3                   1       3   
4                   1       1   
9                   1       3   
10                  1       2   
...               ...     ...   
876                 1       3   
880                 1       1   
881                 1       2   
888                 1       1   
890                 1       1   

                                                          Name     Sex   Age  \
PassengerId                                                                    
2            Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0   
3                                       Heikkinen, Miss. Laina  female  26.0   
4                 Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0   
9            Johnson, Mrs. Oscar W (Elisabeth Vilhelmina Berg)  female  27.0   
10                         Nasser, Mrs. Nicholas (Adele Achem)  female  14.0   
...                                                        ...     ...   ...   
876                           Najib, Miss. Adele Kiamie "Jane"  female  15.0   
880              Potter, Mrs. Thomas Jr (Lily Alexenia Wilson)  female  56.0   
881               Shelley, Mrs. William (Imanita Parrish Hall)  female  25.0   
888                               Graham, Miss. Margaret Edith  female  19.0   
890                                      Behr, Mr. Karl Howell    male  26.0   

             SibSp  Parch            Ticket     Fare Cabin Embarked  
PassengerId                                                          
2                1      0          PC 17599  71.2833   C85        C  
3                0      0  STON/O2. 3101282   7.9250   NaN        S  
4                1      0            113803  53.1000  C123        S  
9                0      2            347742  11.1333   NaN        S  
10               1      0            237736  30.0708   NaN        C  
...            ...    ...               ...      ...   ...      ...  
876              0      0              2667   7.2250   NaN        C  
880              0      1             11767  83.1583   C50        C  
881              0      1            230433  26.0000   NaN        S  
888              0      0            112053  30.0000   B42        S  
890              0      0            111369  30.0000  C148        C  

[342 rows x 11 columns]

In [49]:
df[df.Fare > 50]

Survived  Pclass  \
PassengerId                     
2                   1       1   
4                   1       1   
7                   0       1   
28                  0       1   
32                  1       1   
...               ...     ...   
857                 1       1   
864                 0       3   
868                 0       1   
872                 1       1   
880                 1       1   

                                                          Name     Sex   Age  \
PassengerId                                                                    
2            Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0   
4                 Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0   
7                                      McCarthy, Mr. Timothy J    male  54.0   
28                              Fortune, Mr. Charles Alexander    male  19.0   
32              Spencer, Mrs. William Augustus (Marie Eugenie)  female   NaN   
...                                                        ...     ...   ...   
857                 Wick, Mrs. George Dennick (Mary Hitchcock)  female  45.0   
864                          Sage, Miss. Dorothy Edith "Dolly"  female   NaN   
868                       Roebling, Mr. Washington Augustus II    male  31.0   
872           Beckwith, Mrs. Richard Leonard (Sallie Monypeny)  female  47.0   
880              Potter, Mrs. Thomas Jr (Lily Alexenia Wilson)  female  56.0   

             SibSp  Parch    Ticket      Fare        Cabin Embarked  
PassengerId                                                          
2                1      0  PC 17599   71.2833          C85        C  
4                1      0    113803   53.1000         C123        S  
7                0      0     17463   51.8625          E46        S  
28               3      2     19950  263.0000  C23 C25 C27        S  
32               1      0  PC 17569  146.5208          B78        C  
...            ...    ...       ...       ...          ...      ...  
857              1      1     36928  164.8667          NaN        S  
864              8      2  CA. 2343   69.5500          NaN        S  
868              0      0  PC 17590   50.4958          A24        S  
872              1      1     11751   52.5542          D35        S  
880              0      1     11767   83.1583          C50        C  

[160 rows x 11 columns]

### 3.5.5. Changing the dataframe

We can change one specific value in dataframe, or entire column.

To change specific value in certain location, we can do like this:

In [51]:
df.iloc[3,5]

1

In [52]:
df.iloc[3,5]=0

In [53]:
df.iloc[3,5]

0

We can change entire column using new Series

In [57]:
df2 = df.copy()
df2['Pclass'] = np.array([1]*len(df2))

In [58]:
df2

Survived  Pclass  \
PassengerId                     
1                   0       1   
2                   1       1   
3                   1       1   
4                   1       1   
5                   0       1   
...               ...     ...   
887                 0       1   
888                 1       1   
889                 0       1   
890                 1       1   
891                 0       1   

                                                          Name     Sex   Age  \
PassengerId                                                                    
1                                      Braund, Mr. Owen Harris    male  22.0   
2            Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0   
3                                       Heikkinen, Miss. Laina  female  26.0   
4                 Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0   
5                                     Allen, Mr. William Henry    male  35.0   
...                                                        ...     ...   ...   
887                                      Montvila, Rev. Juozas    male  27.0   
888                               Graham, Miss. Margaret Edith  female  19.0   
889                   Johnston, Miss. Catherine Helen "Carrie"  female   NaN   
890                                      Behr, Mr. Karl Howell    male  26.0   
891                                        Dooley, Mr. Patrick    male  32.0   

             SibSp  Parch            Ticket     Fare Cabin Embarked  
PassengerId                                                          
1                1      0         A/5 21171   7.2500   NaN        S  
2                1      0          PC 17599  71.2833   C85        C  
3                0      0  STON/O2. 3101282   7.9250   NaN        S  
4                0      0            113803  53.1000  C123        S  
5                0      0            373450   8.0500   NaN        S  
...            ...    ...               ...      ...   ...      ...  
887              0      0            211536  13.0000   NaN        S  
888              0      0            112053  30.0000   B42        S  
889              1      2        W./C. 6607  23.4500   NaN        S  
890              0      0            111369  30.0000  C148        C  
891              0      0            370376   7.7500   NaN        Q  

[891 rows x 11 columns]

## 3.6. Missing Value

We will learn how to handle missing values in dataframe

Missing values are represented as `NaN`. In titanic dataset, for example, column Age of PassengerId 889 is `NaN`.

In [60]:
df

Survived  Pclass  \
PassengerId                     
1                   0       3   
2                   1       1   
3                   1       3   
4                   1       1   
5                   0       3   
...               ...     ...   
887                 0       2   
888                 1       1   
889                 0       3   
890                 1       1   
891                 0       3   

                                                          Name     Sex   Age  \
PassengerId                                                                    
1                                      Braund, Mr. Owen Harris    male  22.0   
2            Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0   
3                                       Heikkinen, Miss. Laina  female  26.0   
4                 Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0   
5                                     Allen, Mr. William Henry    male  35.0   
...                                                        ...     ...   ...   
887                                      Montvila, Rev. Juozas    male  27.0   
888                               Graham, Miss. Margaret Edith  female  19.0   
889                   Johnston, Miss. Catherine Helen "Carrie"  female   NaN   
890                                      Behr, Mr. Karl Howell    male  26.0   
891                                        Dooley, Mr. Patrick    male  32.0   

             SibSp  Parch            Ticket     Fare Cabin Embarked  
PassengerId                                                          
1                1      0         A/5 21171   7.2500   NaN        S  
2                1      0          PC 17599  71.2833   C85        C  
3                0      0  STON/O2. 3101282   7.9250   NaN        S  
4                0      0            113803  53.1000  C123        S  
5                0      0            373450   8.0500   NaN        S  
...            ...    ...               ...      ...   ...      ...  
887              0      0            211536  13.0000   NaN        S  
888              0      0            112053  30.0000   B42        S  
889              1      2        W./C. 6607  23.4500   NaN        S  
890              0      0            111369  30.0000  C148        C  
891              0      0            370376   7.7500   NaN        Q  

[891 rows x 11 columns]

### 3.6.1. Check if dataframe contains `NaN`

We can check if the dataframe contains `NaN` using `df.isnull().values.any()`

In [63]:
df.isnull().values.any()

True

We can check it for specific column, too.

In [64]:
df['Age'].isnull().values.any()

True

In [65]:
df['Survived'].isnull().values.any()

False

### 3.6.2. Delete row which contains NaN

We can delete entire row which has one or more `NaN` values using `df.dropna()`.

In [61]:
df.dropna(how='any')

Survived  Pclass  \
PassengerId                     
2                   1       1   
4                   1       1   
7                   0       1   
11                  1       3   
12                  1       1   
...               ...     ...   
872                 1       1   
873                 0       1   
880                 1       1   
888                 1       1   
890                 1       1   

                                                          Name     Sex   Age  \
PassengerId                                                                    
2            Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0   
4                 Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0   
7                                      McCarthy, Mr. Timothy J    male  54.0   
11                             Sandstrom, Miss. Marguerite Rut  female   4.0   
12                                    Bonnell, Miss. Elizabeth  female  58.0   
...                                                        ...     ...   ...   
872           Beckwith, Mrs. Richard Leonard (Sallie Monypeny)  female  47.0   
873                                   Carlsson, Mr. Frans Olof    male  33.0   
880              Potter, Mrs. Thomas Jr (Lily Alexenia Wilson)  female  56.0   
888                               Graham, Miss. Margaret Edith  female  19.0   
890                                      Behr, Mr. Karl Howell    male  26.0   

             SibSp  Parch    Ticket     Fare        Cabin Embarked  
PassengerId                                                         
2                1      0  PC 17599  71.2833          C85        C  
4                0      0    113803  53.1000         C123        S  
7                0      0     17463  51.8625          E46        S  
11               1      1   PP 9549  16.7000           G6        S  
12               0      0    113783  26.5500         C103        S  
...            ...    ...       ...      ...          ...      ...  
872              1      1     11751  52.5542          D35        S  
873              0      0       695   5.0000  B51 B53 B55        S  
880              0      1     11767  83.1583          C50        C  
888              0      0    112053  30.0000          B42        S  
890              0      0    111369  30.0000         C148        C  

[183 rows x 11 columns]

### 3.6.3. Fill row which contains NaN

Or, we can fill `NaN` with certain values with `df.fillna()`.

In [69]:
df.fillna(value=50)

Survived  Pclass  \
PassengerId                     
1                   0       3   
2                   1       1   
3                   1       3   
4                   1       1   
5                   0       3   
...               ...     ...   
887                 0       2   
888                 1       1   
889                 0       3   
890                 1       1   
891                 0       3   

                                                          Name     Sex   Age  \
PassengerId                                                                    
1                                      Braund, Mr. Owen Harris    male  22.0   
2            Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0   
3                                       Heikkinen, Miss. Laina  female  26.0   
4                 Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0   
5                                     Allen, Mr. William Henry    male  35.0   
...                                                        ...     ...   ...   
887                                      Montvila, Rev. Juozas    male  27.0   
888                               Graham, Miss. Margaret Edith  female  19.0   
889                   Johnston, Miss. Catherine Helen "Carrie"  female  50.0   
890                                      Behr, Mr. Karl Howell    male  26.0   
891                                        Dooley, Mr. Patrick    male  32.0   

             SibSp  Parch            Ticket     Fare Cabin Embarked  
PassengerId                                                          
1                1      0         A/5 21171   7.2500    50        S  
2                1      0          PC 17599  71.2833   C85        C  
3                0      0  STON/O2. 3101282   7.9250    50        S  
4                0      0            113803  53.1000  C123        S  
5                0      0            373450   8.0500    50        S  
...            ...    ...               ...      ...   ...      ...  
887              0      0            211536  13.0000    50        S  
888              0      0            112053  30.0000   B42        S  
889              1      2        W./C. 6607  23.4500    50        S  
890              0      0            111369  30.0000  C148        C  
891              0      0            370376   7.7500    50        Q  

[891 rows x 11 columns]

## 3.7. Merge dataframes

We will learn various ways to merge two dataframes.

In [73]:
df_left = pd.DataFrame({'key': ['foo', 'bar'], 'lval': [1, 2]})
df_right = pd.DataFrame({'key': ['foo', 'bar'], 'rval': [4, 5]})

In [74]:
df_left

key  lval
0  foo     1
1  bar     2

In [75]:
df_right

key  rval
0  foo     4
1  bar     5

### 3.7.1. Concatenating two dataframes

We can concatenate two dataframes using `pd.concat()`. It means, we will stack two dataframes. If there is no matching column, it will have value `NaN`.

In [77]:
pd.concat([df_left, df_right])

key  lval  rval
0  foo   1.0   NaN
1  bar   2.0   NaN
0  foo   NaN   4.0
1  bar   NaN   5.0

### 3.7.2. Merging two dataframes

Or, we can merge two dataframes using `pd.merge()`. It merges two dataframes using `key`.

In [79]:
pd.merge(df_left, df_right, on='key')

key  lval  rval
0  foo     1     4
1  bar     2     5

## 3.7. Grouping

We will learn how to group dataframe with some condition using `df.groupby()`.

In [80]:
df

Survived  Pclass  \
PassengerId                     
1                   0       3   
2                   1       1   
3                   1       3   
4                   1       1   
5                   0       3   
...               ...     ...   
887                 0       2   
888                 1       1   
889                 0       3   
890                 1       1   
891                 0       3   

                                                          Name     Sex   Age  \
PassengerId                                                                    
1                                      Braund, Mr. Owen Harris    male  22.0   
2            Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0   
3                                       Heikkinen, Miss. Laina  female  26.0   
4                 Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0   
5                                     Allen, Mr. William Henry    male  35.0   
...                                                        ...     ...   ...   
887                                      Montvila, Rev. Juozas    male  27.0   
888                               Graham, Miss. Margaret Edith  female  19.0   
889                   Johnston, Miss. Catherine Helen "Carrie"  female   NaN   
890                                      Behr, Mr. Karl Howell    male  26.0   
891                                        Dooley, Mr. Patrick    male  32.0   

             SibSp  Parch            Ticket     Fare Cabin Embarked  
PassengerId                                                          
1                1      0         A/5 21171   7.2500   NaN        S  
2                1      0          PC 17599  71.2833   C85        C  
3                0      0  STON/O2. 3101282   7.9250   NaN        S  
4                0      0            113803  53.1000  C123        S  
5                0      0            373450   8.0500   NaN        S  
...            ...    ...               ...      ...   ...      ...  
887              0      0            211536  13.0000   NaN        S  
888              0      0            112053  30.0000   B42        S  
889              1      2        W./C. 6607  23.4500   NaN        S  
890              0      0            111369  30.0000  C148        C  
891              0      0            370376   7.7500   NaN        Q  

[891 rows x 11 columns]

### 3.7.1 Group by column

We can group by certain column and apply some function.

For example, if we want to group `df` with column `Survived` and apply `mean`, we can do as follows:

In [81]:
df.groupby('Survived').mean()

Pclass        Age     SibSp     Parch       Fare
Survived                                                    
0         2.531876  30.626179  0.553734  0.329690  22.117887
1         1.950292  28.343690  0.470760  0.464912  48.395408

### 3.7.2 Group by multiple columns

We can also group by multiple columns.

In [82]:
df.groupby(['Survived', 'Sex']).mean()

Pclass        Age     SibSp     Parch       Fare
Survived Sex                                                       
0        female  2.851852  25.046875  1.209877  1.037037  23.024385
         male    2.476496  31.618056  0.440171  0.207265  21.960993
1        female  1.918455  28.847716  0.510730  0.515021  51.938573
         male    2.018349  27.276022  0.385321  0.357798  40.821484

## 3.8. Exercises

1. read `titanic.csv` file and define it as `df_titanic` (make column `PassengerId` as index).

2. Select the column `Cabin` of `df_titanic`.

3. Select the row `PassengerId=8`

4. Select the row `5-th` row

5. Select rows, which `Sex` is `female`

6. Group `df_titanic` by `Sex` and calculate `mean` of each group